classes/State.php

```diff
@@ -0,0 +1,319 @@
+
+    /**
+     * Int during BackupDb, allowing the script to increent the number of different file names
+     * String during step RestoreDb, which contains the file to process (Data coming from toRestoreQueryList)
+     * @var String|int Contains the SQL progress
+     */
+    private $dbStep = 0;
+
+    /**
+     * Data filled in upgrade warmup, to avoid risky tasks during the process
+     * @var array|null File containing sample files to be deleted
+     */
+    private $removeList = null;
+    /**
+     *
+     * @var string|null File containing files to be upgraded
+     */
+    private $fileToUpgrade = null;
+    /**
+     * @var string|null File containing modules to be upgraded
+     */
+    private $modulesToUpgrade = null;
+
+    /**
+     * installedLanguagesIso is an array of iso_code of each installed languages
+     * @var array
+     */
+    private $installedLanguagesIso = array();
+    /**
+     * modules_addons is an array of array(id_addons => name_module).
+     * @var array
+     */
+    private $modules_addons = array();
+
+    /**
+     * @var bool Determining if all steps went totally successfully
+     */
+    private $warning_exists = false;
+
+    /**
+     * @param array $savedState from another request
+     */
+    public function importFromArray(array $savedState)
+    {
+        foreach($savedState as $name => $value) {
+            if (!empty($value) && property_exists(__CLASS__, $name)) {
+                $this->{$name} = $value;
+            }
+        }
+        return $this;
+    }
+
+    /**
+     * @param string $encodedData
+     */
+    public function importFromEncodedData($encodedData)
+    {
+        $decodedData = json_decode(base64_decode($encodedData), true);
+        return $this->importFromArray($decodedData['nextParams']);
+    }
+
+    /**
+     * @return array of class properties for export
+     */
+    public function export()
+    {
+        return get_object_vars($this);
+    }
+
+    public function initDefault(Upgrader $upgrader, $prodRootDir, $version)
+    {
+        $postData = 'version='.$this->getInstallVersion().'&method=listing&action=native&iso_code=all';
+        $xml_local = $prodRootDir.DIRECTORY_SEPARATOR.'config'.DIRECTORY_SEPARATOR.'xml'.DIRECTORY_SEPARATOR.'modules_native_addons.xml';
+        $xml = $upgrader->getApiAddons($xml_local, $postData, true);
+
+        $modules_addons = array();
+        if (is_object($xml)) {
+            foreach ($xml as $mod) {
+                $modules_addons[(string)$mod->id] = (string)$mod->name;
+            }
+        }
+        $this->setModulesAddons($modules_addons);
+
+        // installedLanguagesIso is used to merge translations files
+        $installedLanguagesIso = array_map(
+            function($v) { return $v['iso_code']; },
+            \Language::getIsoIds(false)
+        );
+        $this->setInstalledLanguagesIso($installedLanguagesIso);
+
+        $rand = dechex(mt_rand(0, min(0xffffffff, mt_getrandmax())));
```

Comment:
Can you explain why you use this function?

Reply:
Nope, this comes directly from the previous version to generate a random value. https://github.com/PrestaShop/autoupgrade/blob/6e863f7b2d196b1128a5a6018056a917242a612e/AdminSelfUpgrade.php#L651